subprojects/base-services/src/test/groovy/org/gradle/internal/operations/MaxWorkersTest.groovy

```diff
@@ -0,0 +1,137 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.operations
+
+import org.gradle.internal.concurrent.DefaultExecutorFactory
+import org.gradle.test.fixtures.concurrent.ConcurrentSpec
+
+class MaxWorkersTest extends ConcurrentSpec {
+
+    def "BuildOperationProcessor operation start blocks when there are no leases available, taken by BuildOperationWorkerRegistry"() {
+        given:
+        def maxWorkers = 1
+        def registry = new DefaultBuildOperationWorkerRegistry(maxWorkers)
+        def processor = new DefaultBuildOperationProcessor(new DefaultBuildOperationQueueFactory(registry), new DefaultExecutorFactory(), maxWorkers)
+        def processorWorker = new DefaultBuildOperationQueueTest.SimpleWorker()
+
+        when:
+        async {
+            start {
+                def cl = registry.operationStart()
+                instant.worker1
+                thread.block()
+                instant.worker1Finished
+                cl.operationFinish()
+            }
+            start {
+                thread.blockUntil.worker1
+                def child2 = registry.operationStart()
+                processor.run(processorWorker, { queue ->
+                    queue.add(new DefaultBuildOperationQueueTest.TestBuildOperation() {
+                        @Override
+                        void run() {
+                            instant.worker2
+                        }
+                    })
+                })
+                child2.operationFinish()
+            }
+        }
+
+        then:
+        instant.worker2 > instant.worker1Finished
+
+        cleanup:
+        registry?.stop()
+    }
+
+    def "BuildOperationWorkerRegistry operation start blocks when there are no leases available, taken by BuildOperationProcessor"() {
+        given:
+        def maxWorkers = 1
+        def registry = new DefaultBuildOperationWorkerRegistry(maxWorkers)
+        def processor = new DefaultBuildOperationProcessor(new DefaultBuildOperationQueueFactory(registry), new DefaultExecutorFactory(), maxWorkers)
+        def processorWorker = new DefaultBuildOperationQueueTest.SimpleWorker()
+
+        when:
+        async {
+            start {
+                def cl = registry.operationStart()
+                processor.run(processorWorker, { queue ->
+                    queue.add(new DefaultBuildOperationQueueTest.TestBuildOperation() {
+                        @Override
+                        void run() {
+                            instant.worker1
+                            thread.block()
+                            instant.worker1Finished
+                        }
+                    })
+                })
+                cl.operationFinish()
+            }
+            start {
+                thread.blockUntil.worker1
```

Comment:
Same exposure here but probably not as bad since worker2 isn't being executed asynchronously.  In any event, simple enough to shore up.